src/lib.rs

```diff
@@ -1,21 +1,392 @@
+
+        for msg in self.receiver.wait() {
+            let msg = msg.map_err(|_| {
+                Error::from_kind(ErrorKind::CouldNotReadValueFromChannel)
+            })?;
+
+            match msg {
+                StarlingMessage::TaskFinished(id) => {
+                    assert!(self.tasks.remove(&id).is_some());
+
+                    let thread_id = id.into();
+                    let join_handle = self.threads
+                        .remove(&thread_id)
+                        .expect("should have a thread join handle for the finished task");
+                    join_handle
+                        .join()
+                        .expect("should join finished task's thread OK");
+
+                    if id == main.id() {
+                        // TODO: notification of shutdown and joining other threads and things.
+                        return Ok(());
+                    }
+                }
+
+                StarlingMessage::TaskErrored(id, error) => {
+                    assert!(self.tasks.remove(&id).is_some());
+
+                    let thread_id = id.into();
+                    let join_handle = self.threads
+                        .remove(&thread_id)
+                        .expect("should have a thread join handle for the errored task");
+                    join_handle
+                        .join()
+                        .expect("should join errored task's thread OK");
+
+                    if id == main.id() {
+                        // TODO: notification of shutdown and joining other threads and things.
+                        return Err(error);
+                    }
+                }
+
+                StarlingMessage::NewTask(_task_handle, _join_handle) => unimplemented!(),
+            }
+        }
+
+        Ok(())
+    }
+}
+
+/// Messages that threads can send to the starling main thread.
+#[derive(Debug)]
+pub(crate) enum StarlingMessage {
+    /// The task on the given thread completed successfully.
+    TaskFinished(task::TaskId),
+
+    /// The task on the given thread failed with the given error.
+    TaskErrored(task::TaskId, Error),
+
+    /// A new child task was created.
+    NewTask(task::TaskId, thread::JoinHandle<()>),
+}
+
+/// A handle to the Starling system.
+///
+/// A `StarlingHandle` is a capability to schedule IO on the event loop, spawn
+/// work in one of the utility thread pools, and communicate with the Starling
+/// main thread. Handles can be cloned and sent across threads, propagating
+/// these capabilities.
+#[derive(Clone)]
+pub(crate) struct StarlingHandle {
+    options: Arc<Options>,
+    sync_io_pool: CpuPool,
+    cpu_pool: CpuPool,
+    sender: Sender<StarlingMessage>,
+}
+
+impl fmt::Debug for StarlingHandle {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "StarlingHandle {{ .. }}")
+    }
+}
+
+impl StarlingHandle {
+    /// Get the `Options` that this starling system was configured with.
```

Comment:
Uber-nit: can you normalize the spelling regarding "Starling" vs "starling"?

Reply:
Good catch ;)